### PR TITLE
[IMP] im_livechat, mail: replace emoji with unicode.

### PR DESCRIPTION
**Pull request by Xavier-Do**
Remove the support to replace emojis by images. Most browser
now support colored emoji or at least black and white emojis.
The emojis list have been updated to ensure a correct support
for most os/browser.
Users have the possibility to install fonts localy if their os does not
have one of the default emoji font installed.
After default os emoji font, one of twemoji, emojione or noto color
will be selected.
The emoji list is now hardcoded in js in order to reduce the number of
records in shortcodes.

Also include a small fix for livechat to display message field when rating
is bad.

Task #36898
PR #23689

Note: some last changes have to be made before merge:
-change images for livechat ratings
-check that all shortcut are consistent with most used websites using emojis
-check the emoji list on multiple os and experiments css behaviour with fonts installed user side.

### CODE REVIEW

File: addons/mail/static/src/js/services/chat_manager.js [hunk 1]
```diff
@@ -1115,20 +1114,93 @@ var ChatManager =  AbstractService.extend({
...
-                    }
-                }
+                self.cannedResponses.push(_.pick(s, ['id', 'source', 'substitution']));
+            });
+            // emojis
+            self.emojiUnicodes={
+                ":)":"😊",
+                ":D":"😃",
+                "xD":"😆",
+                ":-)":"😂",
```
Comment (Elkasitu) [inline addons/mail/static/src/js/services/chat_manager.js]: I suggest `x'D` for this one

File: addons/mail/static/src/js/services/chat_manager.js [hunk 2]
```diff
@@ -1115,20 +1114,93 @@ var ChatManager =  AbstractService.extend({
...
+                "xp":"😝",
+                "o_o":"😳",
+                ":|":"😐",
+                ":/":"😕",
+                ":(":"😞",
+                ":@":"😱",
+                ":O":"😲",
+                ":'o":"😨",  
+                "3:":"😠",
+                "3:)":"😈",
```
Comment (Elkasitu) [inline addons/mail/static/src/js/services/chat_manager.js]: `>:(` and `>:)`

File: addons/mail/static/src/js/services/chat_manager.js (hunk 1 above)
Comment (Yenthe666) [reply addons/mail/static/src/js/services/chat_manager.js]: @Elkasitu oh please no, everybody uses xD

Comment (Elkasitu) [reply addons/mail/static/src/js/services/chat_manager.js]: for the crying one, I mean :innocent:

Comment (Xavier-Do) [reply addons/mail/static/src/js/services/chat_manager.js]: Yes good idea i think the :-) should be an alternative for :)

Comment (Yenthe666) [reply addons/mail/static/src/js/services/chat_manager.js]: Yes, indeed. :)

File: addons/mail/static/src/js/services/chat_manager.js (hunk 2 above)
Comment (Xavier-Do) [reply addons/mail/static/src/js/services/chat_manager.js]: As an alternative solution yes but 3:) is the facebook shortcut for the devil.

Comment (Elkasitu) [reply addons/mail/static/src/js/services/chat_manager.js]: Damn Zuckerberg!! :angry:

Comment (Xavier-Do) [reply addons/mail/static/src/js/services/chat_manager.js]: As far as I remember it was the same on msn messenger :D

Comment (Yenthe666) [reply addons/mail/static/src/js/services/chat_manager.js]: MSN, wow where have the times gone too. Now you made me nostalgic! 3:

File: addons/im_livechat/static/src/js/im_livechat.js [hunk 3]
```diff
@@ -343,7 +343,13 @@ var Feedback = Widget.extend({
         };
         return session.rpc('/im_livechat/feedback', args).then(function () {
             if (options.close) {
-                var content = _.str.sprintf(_t("Rating: :rating_%d"), self.rating);
+                var rating_to_emoji = {
```
Comment (ged-odoo) [inline addons/im_livechat/static/src/js/im_livechat.js]: you could add this on top of the file:

`var RATING_TO_EMOJI = ...`

File: addons/mail/static/src/js/services/chat_manager.js [hunk 4]
```diff
@@ -1115,19 +1114,97 @@ var ChatManager =  AbstractService.extend({
             self.mentionPartnerSuggestions = result.mention_partner_suggestions;
             self.discussMenuID = result.menu_id;
 
-            // Shortcodes: canned responses and emojis
+            // Shortcodes: canned responses
```
Comment (ged-odoo) [inline addons/mail/static/src/js/services/chat_manager.js]: trailing space...

File: addons/mail/static/src/js/services/chat_manager.js [hunk 5]
```diff
@@ -1115,19 +1114,97 @@ var ChatManager =  AbstractService.extend({
...
-                    self.emojiSubstitutions[_.escape(s.source)] = s.substitution;
-                    if (s.unicode_source) {
-                        self.emojiSubstitutions[_.escape(s.unicode_source)] = s.substitution;
-                        self.emojiUnicodes[_.escape(s.source)] = s.unicode_source;
-                    }
+                self.cannedResponses.push(_.pick(s, ['id', 'source', 'substitution']));
+            });
+            // emojis
+
+            var lastadded = null;
```
Comment (ged-odoo) [inline addons/mail/static/src/js/services/chat_manager.js]: `var lastAdded = null;`

File: addons/mail/static/src/js/services/chat_manager.js [hunk 6]
```diff
@@ -1115,19 +1114,97 @@ var ChatManager =  AbstractService.extend({
...
+                ":guitar":"🎸",
+                ":soccer":"⚽️",
+                ":football":"🏈",
+                ":8ball":"🎱",
+                ":clapper":"🎬",
+                ":microphone":"🎤"
+                };
+            _.each(_.keys(self.emojiUnicodes), function (key) {
+                var unicode = self.emojiUnicodes[key];
+                if (lastadded != unicode){
```
Comment (ged-odoo) [inline addons/mail/static/src/js/services/chat_manager.js]: `if (lastAdded !== unicode) {`

File: addons/mail/static/src/js/services/chat_manager.js [hunk 7]
```diff
@@ -1115,19 +1114,97 @@ var ChatManager =  AbstractService.extend({
...
+                ":music":"🎵",
+                ":trumpet":"🎺",
+                ":guitar":"🎸",
+                ":soccer":"⚽️",
+                ":football":"🏈",
+                ":8ball":"🎱",
+                ":clapper":"🎬",
+                ":microphone":"🎤"
+                };
+            _.each(_.keys(self.emojiUnicodes), function (key) {
```
Comment (ged-odoo) [inline addons/mail/static/src/js/services/chat_manager.js]: `_.each(self.emojiUnicodes, function (unicode, key) {`

_.each can be used on objects

File: addons/mail/static/src/js/services/chat_manager.js [hunk 8]
```diff
@@ -1275,13 +1352,12 @@ var ChatManager =  AbstractService.extend({
             url: session.url("/mail/view?message_id=" + data.id),
             module_icon:data.module_icon,
         };
-
-        _.each(_.keys(this.emojiSubstitutions), function (key) {
-            var escapedKey = String(key).replace(/([.*+?=^!:${}()|[\]/\\])/g, '\\$1');
-            var regexp = new RegExp("(?:^|\\s|<[a-z]*>)(" + escapedKey + ")(?=\\s|$|</[a-z]*>)", "g");
-            msg.body = msg.body.replace(regexp, ' <span class="o_mail_emoji">'+self.emojiSubstitutions[key]+'</span> ');
+        _.each(_.keys(self.emojiUnicodes), function (key) {
```
Comment (ged-odoo) [inline addons/mail/static/src/js/services/chat_manager.js]: `_.each(self.emojiUnicodes, function (value, key) {`

File: addons/mail/static/src/js/services/chat_manager.js [hunk 9]
```diff
@@ -1275,13 +1352,12 @@ var ChatManager =  AbstractService.extend({
...
+            //add o_mail_emoji class on each unicode to manage size and font
+            var unicode = String(self.emojiUnicodes[key]);
```
Comment (ged-odoo) [inline addons/mail/static/src/js/services/chat_manager.js]: var unicode = String(value);

File: addons/mail/static/src/js/services/chat_manager.js [hunk 10]
```diff
@@ -1115,19 +1114,97 @@ var ChatManager =  AbstractService.extend({
...
-                    if (s.unicode_source) {
-                        self.emojiSubstitutions[_.escape(s.unicode_source)] = s.substitution;
-                        self.emojiUnicodes[_.escape(s.source)] = s.unicode_source;
-                    }
+                self.cannedResponses.push(_.pick(s, ['id', 'source', 'substitution']));
+            });
+            // emojis
+
+            var lastadded = null;
+            self.emojiUnicodes={
```
Comment (ged-odoo) [inline addons/mail/static/src/js/services/chat_manager.js]: i don't think this needs to be a class variable.  I think you should put this on top of the file as a constant.

File: addons/mail/static/src/js/services/chat_manager.js (hunk 4 above)
Comment (Xavier-Do) [reply addons/mail/static/src/js/services/chat_manager.js]: So close

File: addons/delivery/views/product_packaging_view.xml [hunk 11]
```diff
@@ -1,47 +1,37 @@
 <?xml version="1.0" encoding="utf-8"?>
 <odoo>
 
-<record model="ir.ui.view" id="product_packaging_delivery_form">
+<record model="ir.ui.view" id="delivery.product_packaging_delivery_form">
```
Comment (tde-banana-odoo) [inline addons/delivery/views/product_packaging_view.xml]: I bet you have had some issues in a rebase. I don't think you had to modify delivery, am I wrong ?

File: addons/im_livechat/__manifest__.py [hunk 12]
```diff
@@ -26,8 +26,7 @@
         "views/im_livechat_channel_views.xml",
         "views/im_livechat_channel_templates.xml",
         "report/im_livechat_report_channel_views.xml",
-        "report/im_livechat_report_operator_views.xml",
-        "data/im_livechat_channel_data.xml"
+        "report/im_livechat_report_operator_views.xml"
```
Comment (tde-banana-odoo) [inline addons/im_livechat/__manifest__.py]: You removed im_livechat_channel_data import ? You modified the file but also removed its import it seems.

File: addons/im_livechat/static/src/js/im_livechat.js (hunk 3 above)
Comment (tde-banana-odoo) [reply addons/im_livechat/static/src/js/im_livechat.js]: Small up on this comment, did you handle it ?

**Pull request merged**